### ISSUE TITLE
TopK for Tensor?

### ISSUE THREAD

**Issue by zzsfornlp** [clab/dynet#1263]
Faster TopK functions for Tensor can be useful for processes like beam search. Correct me if I'm wrong, currently I have not found efficient way to do this directly on GPU in dynet.

It seems that tensorflow and pytorch adopt two different methods for this (but seemingly not with Eigen), which can be good references or resources:
Tensorflow (heap-based):
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/topk_op_gpu.cu.cc
PyTorch (radix-selection):
https://github.com/pytorch/pytorch/blob/master/aten/src/THC/THCTensorTopK.cuh

Will it be ok to adapt them into dynet? (Putting the relevant codes in `third_party` with corresponding comments about lisence and sources.)
Maybe something like:
`pair<IndexTensor, Tensor> TensorTools::topk(const Tensor& v, unsigned dim = 0, unsigned k = 1);`
If ok, I would like to work on this.

**Comment by neubig** [comment]
Yes, it's OK to adopt code form other Apache licensed projects to DyNet. If the license is something other than Apache, we'll need to be careful.

We welcome a pull request on this.
